pipeline/backend/ssh/ssh.go

```diff
@@ -0,0 +1,141 @@
+package ssh
+
+import (
+	"context"
+	"encoding/base64"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/melbahja/goph"
+
+	"github.com/woodpecker-ci/woodpecker/pipeline/backend/types"
+	"github.com/woodpecker-ci/woodpecker/server"
+)
+
+type ssh struct {
+	cmd    *goph.Cmd
+	output io.ReadCloser
+	client *goph.Client
+}
+
+type readCloser struct {
+	io.Reader
+}
+
+func (c readCloser) Close() error {
+	return nil
+}
+
+// make sure local implements Engine
+var _ types.Engine = &ssh{}
+
+// New returns a new ssh Engine.
+func New() types.Engine {
+	return &ssh{}
+}
+
+func (e *ssh) Name() string {
+	return "ssh"
+}
+
+func (e *ssh) IsAvailable() bool {
+	return true
+}
+
+func (e *ssh) Load() error {
+	var auth goph.Auth
+	if file, has := os.LookupEnv("WOODPECKER_SSH_KEY"); has {
+		var err error
+		auth, err = goph.Key(file, os.Getenv("WOODPECKER_SSH_KEY_PASSWORD"))
+		if err != nil {
+			return err
+		}
+	} else {
+		auth = goph.Password(os.Getenv("WOODPECKER_SSH_PASSWORD"))
+	}
+	client, err := goph.New(os.Getenv("WOODPECKER_SSH_USER"), os.Getenv("WOODPECKER_SSH_HOST"), auth)
+	if err != nil {
+		return err
+	}
+	e.client = client
+	return nil
+}
+
+// Setup the pipeline environment.
+func (e *ssh) Setup(ctx context.Context, proc *types.Config) error {
+	return nil
+}
+
+// Exec the pipeline step.
+func (e *ssh) Exec(ctx context.Context, proc *types.Step) error {
+	// Get environment variables
+	Command := []string{}
+	for a, b := range proc.Environment {
+		if a != "HOME" && a != "SHELL" { // Don't override $HOME and $SHELL
+			Command = append(Command, a+"="+b)
+		}
+	}
+
+	// Get default clone image
+	defaultCloneImage := "docker.io/woodpeckerci/plugin-git:latest"
+	if len(server.Config.Pipeline.DefaultCloneImage) > 0 {
+		defaultCloneImage = server.Config.Pipeline.DefaultCloneImage
+	}
+
+	if proc.Image == defaultCloneImage {
+		// Default clone step
+		Command = append(Command, "CI_WORKSPACE=/tmp/woodpecker/"+proc.Environment["CI_REPO"])
```

Comment:
Using /tmp with a predictable name is a bad idea. If someone has a access to the server, that person could simply inject content in /tmp in various fashion, which in turn mean code execution under the user that is running the CI (and or inject content if the CI is used to make artifact).

It would also fail if 2 jobs are running at the same time, and trust me, that's always extra fun to debug.

Reply:
That part is explained on #872

Reply:
I ported the changes to the `ssh` backend.